src/zarr/codecs/sharding.py

```diff
@@ -473,6 +473,10 @@ async def _decode_partial_single(
             shard_dict = {}
             for chunk_coords in all_chunk_coords:
                 chunk_byte_slice = shard_index.get_chunk_slice(chunk_coords)
+                # the byte_getter wants the start index and total length to read
+                # but chunk_byte_slice is the start and end index
+                chunk_byte_slice = (chunk_byte_slice[0],
```

Comment:
`chunk_byte_slice` can return `None`, and thus its return value cannot be indexed without first narrowing the type. @normanrz any guidance for how we should handle `None` here?

Reply:
The downstream byte getter accepts `None` in which case it reads the whole shard. So I could pass `None` through as is, but that's probably an error case here since we're supposed to be partially reading a shard.

Reply:
I guess the easiest thing here is to do 

```
if chunk_byte_slice is not None: 
  # your indexing fix
return chunk_byte_slice
 ```
since the current behavior already propagates `None`

Reply:
I think we should refactor the `_ShardIndex` to return `(index, offset) | None`. That is the internal format of the shard index, anyways.

Reply:
Okay the latest commit implements the suggested refactor.